code/game/machinery/sorter.dm

```diff
@@ -0,0 +1,264 @@
+	var/refuse_output_side = null		//by default it will be reversed sorter's dir
+
+	var/progress = 0
+
+	var/list/sort_settings = list()
+	var/obj/item/current_item
+
+	//UI vars
+	var/list/custom_rule = list("accept", "sort_type", "value", "amount")
+	var/new_rule_ui = FALSE
+
+
+/obj/machinery/sorter/Destroy()
+	if(current_item)
+		current_item.forceMove(get_turf(src))
+	return ..()
+
+
+/obj/machinery/sorter/update_icon()
+	..()
+	if(progress)
+		icon_state = "sorter-process"
+	else
+		icon_state = "sorter"
+
+
+/obj/machinery/sorter/Process()
+	if(stat & BROKEN || stat & NOPOWER)
+		progress = 0
+		use_power(0)
+		update_icon()
+		return
+
+	if(!sort_settings)
+		return
+
+	if(current_item)
+		use_power(2)
+		progress += speed
+		if(progress >= 100)
+			sort(current_item)
+			progress = 0
+			use_power(1)
+		update_icon()
+	else
+		grab()
+
+
+/obj/machinery/sorter/proc/sort(var/obj/item_to_sort)
+	if(!current_item || !sort_settings)
+		return
+	var/sorted = FALSE
+	for(var/datum/sort_rule/rule in sort_settings)
+		switch(rule.sort_type)
+			if(SORT_TYPE_MATERIAL)
+				if(rule.value in item_to_sort.matter)
+					if(rule.amount && (item_to_sort.matter[rule.value] >= rule.amount))
+						sorted = rule.accept
+						if(!sorted)
+							break
+
+			if(SORT_TYPE_REAGENT)
+				if(item_to_sort.reagents && item_to_sort.reagents.has_reagent(rule.value, rule.amount))
+					sorted = rule.accept
+					if(!sorted)
+						break
+
+	eject(sorted)
+	return
+
+
+/obj/machinery/sorter/proc/grab()
+	if(current_item)
+		return
+	var/turf/T = get_step(src, dir)
+	var/obj/item/O = locate(/obj/item) in T
+	if(istype(O) && !O.anchored)
+		current_item = O
+		O.forceMove(src)
+		state("scanning now: [O]...")
+
+
+/obj/machinery/sorter/proc/eject(var/sorted = FALSE)
+	if(!current_item)
+		return
+	var/output_dir
+	if(refuse_output_side)
+		output_dir = refuse_output_side
+	else
+		output_dir = reverse_direction(dir)
+	var/turf/T
+	if(sorted)
+		T = get_step(src, accept_output_side)
+		state("[current_item] accepted.")
+	else
+		T = get_step(src, output_dir)
+		state("[current_item] refused.")
+	if(T)
+		current_item.forceMove(T)
+		current_item = null
+
+
+/obj/machinery/sorter/RefreshParts()
+	..()
+	var/manipulator_rating = 0
+	for(var/obj/item/weapon/stock_parts/manipulator/M in component_parts)
+		manipulator_rating += M.rating
+	var/num_settings = 0
+	for(var/obj/item/weapon/stock_parts/scanning_module/S in component_parts)
+		num_settings += S.rating
+	number_of_settings = num_settings * 2
+	speed = manipulator_rating*4
+
+
+/obj/machinery/sorter/attackby(var/obj/item/I, var/mob/user)
+	if(default_deconstruction(I, user))
+		return
+
+	if(default_part_replacement(I, user))
+		return
+
+	..()
+
+/obj/machinery/sorter/attack_hand(mob/user as mob)
+	return ui_interact(user)
+
+
+//UI
+/obj/machinery/sorter/ui_interact(mob/user, ui_key = "main", var/datum/nanoui/ui = null, var/force_open = 1, var/datum/topic_state/state = GLOB.default_state)
```

Comment:
PBJ did nano_ui improve PR recently and added new argument here, please look for it

Reply:
Sure